tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/client/RollbackTest.java

```diff
@@ -0,0 +1,116 @@
+import org.apache.activemq.artemis.spi.core.protocol.RemotingConnection;
+import org.apache.activemq.artemis.tests.util.ActiveMQTestBase;
+import org.jboss.logging.Logger;
+import org.junit.Before;
+import org.junit.Test;
+
+public class RollbackTest extends ActiveMQTestBase {
+
+   protected ActiveMQServer server;
+
+   private static Logger log = Logger.getLogger(RollbackTest.class);
+
+   @Override
+   @Before
+   public void setUp() throws Exception {
+      super.setUp();
+      server = createServer(false, createDefaultInVMConfig());
+      server.getConfiguration().getIncomingInterceptorClassNames().add(MyInterceptor.class.getName());
+      server.start();
+   }
+
+   @Test
+   public void testFailedRollback() throws Exception {
+      final String TOPIC = "myTopic";
+      final String SUBSCRIPTION = "mySub";
+
+      try (ActiveMQConnectionFactory connectionFactory = new ActiveMQConnectionFactory("vm://0")) {
+         connectionFactory.setCallTimeout(1000); // fail fast
+         connectionFactory.setReconnectAttempts(-1);
+         connectionFactory.setConfirmationWindowSize(1024 * 1024);
+
+         try (Connection consumerConnection = connectionFactory.createConnection()) {
+            consumerConnection.start();
+            final Session session = consumerConnection.createSession(Session.SESSION_TRANSACTED);
+            Topic topic = session.createTopic(TOPIC);
+            final MessageConsumer messageConsumer = session.createSharedDurableConsumer(topic, SUBSCRIPTION);
+            MessageProducer p = session.createProducer(topic);
+            p.send(session.createMessage());
+            p.close();
+            session.commit();
+
+            try {
+               Message m = messageConsumer.receive(2000);
+               assertNotNull(m);
+               // the interceptor will block this first rollback and trigger a failure, the failure will cause the client to re-attach its session
+               session.rollback();
+               fail();
+            } catch (JMSException jmsException) {
+               // expected
+            }
+
+            try {
+               session.rollback();
+            } catch (JMSException e) {
+               fail("Rollback failed again! Giving up. " + e.getMessage());
+            }
+
+            Message m = messageConsumer.receive(2000);
+            assertNotNull(m);
+            try {
+               session.commit();
+            } catch (JMSException e) {
+               fail("Commit failed. " + e.getMessage());
+            }
+         }
+      }
+
+      assertEquals(0L, server.locateQueue(SUBSCRIPTION).getMessageCount());
+   }
+
+   public static class MyInterceptor implements Interceptor {
+      private boolean intercepted = false;
+
+      @Override
+      public boolean intercept(Packet packet, RemotingConnection connection) throws ActiveMQException {
+         if (!intercepted && packet.getType() == PacketImpl.SESS_ROLLBACK) {
```

Comment:
you need to take the rollback out of the communication? is this failover or regular operation?

Reply:
(this was not a review of course.. I'm just trying to understand the issue).

Reply:
This is just for the situation where the rollback fails (e.g. spurious network issue) and the client automatically reconnects.